### DISCUSSION TITLE
Starting Designer on Mac still a hard problem due to JAVAFX

### DISCUSSION THREAD

**Discussion by rsoesemann** [pmd/pmd#4153] [Q&A] [answered]
_This seems to be an enduring issue that I have every few months when I want to use the PMD Designer. No existing documentation or community content online seems to cover or solve my case._

When I 

```
export PATH_TO_FX=/Users/rsoesemann/dev/javafx-sdk-11.0.2/lib 
export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-11.0.2/lib 
./run.sh designer
```
I get

> You seem to be missing the JavaFX runtime.
>  Please install JavaFX on your system and try again.
>  See https://gluonhq.com/products/javafx/

I am using PMD 6.50 on Mac OS Monterey. I did not install FX just downloaded the SDK from https://download2.gluonhq.com/openjfx/11.0.2/openjfx-11.0.2_osx-x64_bin-sdk.zip. But something seems missing at least calling

> JavaFX -version 

reveals 

> zsh: command not found: javafx

**Comment by jsotuyod** [comment, marked as answer]
You should not include the /lib in the env var. It should be:

```
export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-11.0.2/
./run.sh designer
```

**Reply by rsoesemann** [reply]
It then fails with:

```
rsoesemann@MacBook-Pro-von-Robert ~ % cd dev/pmd-bin-6.50.0/bin 
rsoesemann@MacBook-Pro-von-Robert bin % export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-11.0.2/
rsoesemann@MacBook-Pro-von-Robert bin % ./run.sh designer
Exception in thread "main" java.lang.NullPointerException: inStream parameter is null
	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
	at java.base/java.util.Properties.load(Properties.java:407)
	at net.sourceforge.pmd.util.fxdesigner.DesignerStarter.getJavaFxVersion(DesignerStarter.java:116)
	at net.sourceforge.pmd.util.fxdesigner.DesignerStarter.isCompatibleJavaFxVersion(DesignerStarter.java:100)
	at net.sourceforge.pmd.util.fxdesigner.DesignerStarter.launchGui(DesignerStarter.java:153)
	at net.sourceforge.pmd.util.fxdesigner.DesignerStarter.main(DesignerStarter.java:86)
rsoesemann@MacBook-Pro-von-Robert bin % 
```

**Reply by rsoesemann** [reply]
Looks like it fails to read a property file.

**Reply by jsotuyod** [reply]
That I fear is a bug I recently found myself. At some point JavaFX changed the location of a properties file from being within a jar to laying bare on the lib dir. Until we ship a new version, you can manually patch your run.sh script this way https://github.com/pmd/pmd/blob/ed9bfc952049f489a91617187f54e5edb6b04dce/pmd-dist/src/main/resources/scripts/pmd#L168-L174

Sorry for the inconvenience.

**Reply by rsoesemann** [reply]
I guess I patched incorrectly but I get

```
rsoesemann@MacBook-Pro-von-Robert bin % export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-11.0.2/
rsoesemann@MacBook-Pro-von-Robert bin % ./run.sh designer        
You seem to be running an older version of JavaFX runtime.
 Please install the latest JavaFX on your system and try again.
 See https://gluonhq.com/products/javafx/
```

when I added the yellow part, which results in 

```
#!/bin/bash

usage() {
    echo "Usage:"
    echo "    $(basename "$0") <application-name> [-h|-v] ..."
    echo ""
    echo "application-name: valid options are: $(valid_app_options)"
    echo "-h print this help"
    echo "-v display PMD's version"
}

valid_app_options () {
    echo "pmd, cpd, cpdgui, designer, bgastviewer, designerold, ast-dump"
}

is_cygwin() {
    case "$(uname)" in
        CYGWIN*|MINGW*)
            readonly cygwin=true
            ;;
    esac
    # OS specific support.  $var _must_ be set to either true or false.
    if [ -z ${cygwin} ] ; then
        readonly cygwin=false
    fi
}

cygwin_paths() {
    # For Cygwin, switch paths to Windows format before running java
    if ${cygwin} ; then
        [ -n "${JAVA_HOME}" ] && JAVA_HOME=$(cygpath --windows "${JAVA_HOME}")
        [ -n "${JAVAFX_HOME}" ] && JAVAFX_HOME=$(cygpath --windows "${JAVAFX_HOME}")
        [ -n "${DIRECTORY}" ] && DIRECTORY=$(cygpath --windows "${DIRECTORY}")
        classpath=$(cygpath --path --windows "${classpath}")
    fi
}

convert_cygwin_vars() {
    # If cygwin, convert to Unix form before manipulating
    if ${cygwin} ; then
        [ -n "${JAVA_HOME}" ] && JAVA_HOME=$(cygpath --unix "${JAVA_HOME}")
        [ -n "${JAVAFX_HOME}" ] && JAVAFX_HOME=$(cygpath --unix "${JAVAFX_HOME}")
        [ -n "${CLASSPATH}" ] && CLASSPATH=$(cygpath --path --unix "${CLASSPATH}")
    fi
}

java_heapsize_settings() {
    local heapsize=${HEAPSIZE}
    case "${heapsize}" in
        [1-9]*[mgMG])
            readonly HEAPSIZE="-Xmx${heapsize}"
            ;;
        '')
            ;;
        *)
            echo "HEAPSIZE '${HEAPSIZE}' unknown (try: 1024m)"
            exit 1
    esac
}


set_lib_dir() {
  if [ -z "${LIB_DIR}" ]; then
    # Allow for symlinks to this script
    if [ -L "$0" ]; then
      local script_real_loc=$(readlink "$0")
    else
      local script_real_loc=$0
    fi
    local script_dir=$(dirname "${script_real_loc}")
    local cwd="${PWD}"

    cd "${script_dir}/../lib"
    readonly LIB_DIR=$(pwd -P)
    cd "${cwd}"
  fi
}

check_lib_dir() {
  if [ ! -e "${LIB_DIR}" ]; then
    echo "The jar directory [${LIB_DIR}] does not exist"
  fi
}

function script_exit() {
    echo "$1" >&2
    exit 1
}

determine_java_version() {
    local full_ver=$(java -version 2>&1)
    # java_ver is eg "80" for java 1.8, "90" for java 9.0, "100" for java 10.0.x
    readonly java_ver=$(echo "$full_ver" | sed -n '{
        # replace early access versions, e.g. 11-ea with 11.0.0
        s/-ea/.0.0/
        # replace versions such as 10 with 10.0.0
        s/version "\([0-9]\{1,\}\)"/version "\1.0.0"/
        # replace old java versions 1.x.* (java 1.7, java 1.8) with x.*
        s/version "1\.\(.*\)"/version "\1"/
        # extract the major and minor parts of the version
        s/^.* version "\([0-9]\{1,\}\)\.\([0-9]\{1,\}\).*".*$/\1\2/p
    }')
    # java_vendor is either java (oracle) or openjdk
    readonly java_vendor=$(echo "$full_ver" | sed -n -e 's/^\(.*\) version .*$/\1/p')
}

jre_specific_vm_options() {
  if [ "${APPNAME}" = "designer" ]
  then
    options=""

    if [ "$java_ver" -ge 80 ] && [ "$java_ver" -lt 90 ]
    then
      # no options needed for java8.
      options=""
    elif [ "$java_ver" -ge 90 ] && [ "$java_ver" -lt 110 ] && [ "$java_vendor" = "java" ]
    then
      # java9 and java10 from oracle contain javafx as a module
      # open internal module of javafx to reflection (for our TreeViewWrapper)
      options="--add-opens javafx.controls/javafx.scene.control.skin=ALL-UNNAMED"
      # The rest here is for RichtextFX
      options+=" --add-opens javafx.graphics/javafx.scene.text=ALL-UNNAMED"
      options+=" --add-opens javafx.graphics/com.sun.javafx.scene.text=ALL-UNNAMED"
      options+=" --add-opens javafx.graphics/com.sun.javafx.text=ALL-UNNAMED"
      options+=" --add-opens javafx.graphics/com.sun.javafx.geom=ALL-UNNAMED"
      # Warn of remaining illegal accesses
      options+=" --illegal-access=warn"
    elif [ "$java_vendor" = "openjdk" ] || ( [ "$java_vendor" = "java" ] && [ "$java_ver" -ge 110 ] )
    then
      # openjdk and java11 from oracle onwards do not contain javafx directly
      # there are no extra options either - javafx will be added to the classpath without modules
      options=""
    fi

    echo $options
  else
    echo ""
  fi
}

function add_pmd_classpath() {
    if [ -n "$classpath" ]; then
        classpath="$classpath:${LIB_DIR}/*"
    else
        classpath="${LIB_DIR}/*"
    fi
}

function add_openjfx_classpath() {
  if [ "${APPNAME}" = "designer" ]
  then
    if [ "$java_vendor" = "openjdk" ] && [ "$java_ver" -lt 100 ]
    then
      script_exit "For openjfx at least java 10 is required"
    elif [ "$java_vendor" = "openjdk" ] || ( [ "$java_vendor" = "java" ] && [ "$java_ver" -ge 110 ] )
    then
      # openjfx is required for openjdk builds and oracle java 11 or later
      if [ -z "${JAVAFX_HOME}" ]
      then
        script_exit "The environment variable JAVAFX_HOME is missing."
      else
        # The wildcard will include only jar files, but we need to access also
        # property files such as javafx.properties that lay bare in the dir
        if [ -n "$classpath" ]; then
          classpath="$classpath:${JAVAFX_HOME}/lib/*:${JAVAFX_HOME}/lib/"
        else
          classpath="${JAVAFX_HOME}/lib/*:${JAVAFX_HOME}/lib/"
        fi
      fi
    fi
  fi
}


readonly APPNAME="${1}"
if [ -z "${APPNAME}" ]; then
    usage
    exit 1
fi
shift

case "${APPNAME}" in
  "pmd")
    readonly CLASSNAME="net.sourceforge.pmd.PMD"
    ;;
  "cpd")
    readonly CLASSNAME="net.sourceforge.pmd.cpd.CPD"
    ;;
  "designer")
    readonly CLASSNAME="net.sourceforge.pmd.util.fxdesigner.DesignerStarter"
    ;;
  "designerold")
    readonly CLASSNAME="net.sourceforge.pmd.util.designer.Designer"
    ;;
  "bgastviewer")
    readonly CLASSNAME="net.sourceforge.pmd.util.viewer.Viewer"
    ;;
  "cpdgui")
    readonly CLASSNAME="net.sourceforge.pmd.cpd.GUI"
    ;;
  "ast-dump")
    readonly CLASSNAME="net.sourceforge.pmd.util.treeexport.TreeExportCli"
    ;;
  *)
    echo "${APPNAME} is NOT a valid application name, valid options are: $(valid_app_options)"
    ;;
esac

is_cygwin

set_lib_dir
check_lib_dir

convert_cygwin_vars

classpath=$CLASSPATH

add_pmd_classpath
determine_java_version
add_openjfx_classpath

cygwin_paths

java_heapsize_settings

java ${HEAPSIZE} ${PMD_JAVA_OPTS} $(jre_specific_vm_options) -cp "${classpath}" "${CLASSNAME}" "$@"
```

**Reply by rsoesemann** [reply]
@jsotuyod I switched to a v19 of JavaFX and get this

```
rsoesemann@MacBook-Pro-von-Robert bin % export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-19    
rsoesemann@MacBook-Pro-von-Robert bin % ./run.sh designer
Oct 13, 2022 1:44:30 PM com.sun.javafx.application.PlatformImpl startup
WARNING: Unsupported JavaFX configuration: classes were loaded from 'unnamed module @50aac7a1'
Loading library prism_es2 from resource failed: java.lang.UnsatisfiedLinkError: /Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib: dlopen(/Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib, 0x0001): tried: '/Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))
java.lang.UnsatisfiedLinkError: /Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib: dlopen(/Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib, 0x0001): tried: '/Users/rsoesemann/.openjfx/cache/19+11/x86_64/libprism_es2.dylib' (mach-o file, but is an incompatible architecture (have (arm64), need (x86_64)))
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2430)
```

**Reply by jsotuyod** [reply]
You patched the script just fine. The properties file is read and used to determine the installed version of JavaFX. We are now requiring JavaFX 14+ in MacOS to avoid it from rendering garbled text (see https://github.com/pmd/pmd-designer/issues/10#issuecomment-741786111). Updating to JavaFX 19 is the right move.

This new error may be that you downloaded the wrong arch version (downloaded aarch64, but your machine is x86_64), is that so?

If not, does your machine have an M1 chip? Do you have Rossetta Apple installed? unfortunately, I do not own an M1 Mac, so I haven't tested this, but I know some apps do have challenges running on it…

**Reply by rsoesemann** [reply]
@jsotuyod Replaced it with JavaFX 14 and now I receive a Mac Error
![Screenshot 2022-10-13 at 14 15 40](https://user-images.githubusercontent.com/8180281/195594040-b733f77e-6497-4d9c-aef7-ac0e73032a48.png)

“libprism_es2.dylib” cannot be opened because the developer cannot be verified.

**Reply by rsoesemann** [reply]
It seems I force Mac to allow it but now I see

.```
/run.sh designer        
Exception in Application start method
java.lang.RuntimeException: Exception in Application start method
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:900)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.UnsatisfiedLinkError: no javafx_iio in java.library.path: [/Users/rsoesemann/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2660)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:827)
	at java.base/java.lang.System.loadLibrary(System.java:1871)
	at com.sun.glass.utils.NativeLibLoader.loadLibraryInternal(NativeLibLoader.java:163)
	at com.sun.glass.utils.NativeLibLoader.loadLibrary(NativeLibLoader.java:53)
	at com.sun.javafx.iio.jpeg.JPEGImageLoader.lambda$static$0(JPEGImageLoader.java:107)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.iio.jpeg.JPEGImageLoader.<clinit>(JPEGImageLoader.java:106)
	at com.sun.javafx.iio.jpeg.JPEGImageLoaderFactory.createImageLoader(JPEGImageLoaderFactory.java:49)
	at com.sun.javafx.iio.ImageStorage.getLoaderBySignature(ImageStorage.java:429)
	at com.sun.javafx.iio.ImageStorage.loadAll(ImageStorage.java:268)
	at com.sun.javafx.tk.quantum.PrismImageLoader2.loadAll(PrismImageLoader2.java:142)
	at com.sun.javafx.tk.quantum.PrismImageLoader2.<init>(PrismImageLoader2.java:77)
	at com.sun.javafx.tk.quantum.QuantumToolkit.loadImage(QuantumToolkit.java:794)
	at javafx.scene.image.Image.loadImage(Image.java:1078)
	at javafx.scene.image.Image.initialize(Image.java:813)
	at javafx.scene.image.Image.<init>(Image.java:703)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at net.sourceforge.pmd.util.fxdesigner.Designer.setIcons(Designer.java:148)
	at net.sourceforge.pmd.util.fxdesigner.Designer.start(Designer.java:78)
	at net.sourceforge.pmd.util.fxdesigner.Designer.start(Designer.java:65)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:846)
	at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:455)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
```

**Reply by jsotuyod** [reply]
`libjavafx_iio.dylib` should be part of JavaFX lib directory… make sure it's there (and you allowed it to be opened / is not an untrusted source).

**Reply by rsoesemann** [reply]
I think that part is fixed now as I am seeing this other stacktracy error

**Reply by rsoesemann** [reply]
No @jsotuyod you were right. The Mac warning put 2 of the libs in the recycle bin. I put them back where they belong and now it works.

**Reply by jsotuyod** [reply]
Glad to hear that!

**Reply by rsoesemann** [reply]
### Summary

1. Download JavaFX 14 from https://download2.gluonhq.com/openjfx/14/openjfx-14_osx-x64_bin-sdk.zip
2. Make sure you choose the SDK for your Mac architecture
3. Unzip it and put it in a folder under /Users
4. Open a Terminal window
5.` export JAVAFX_HOME=/Users/rsoesemann/dev/javafx-sdk-14`
6. `./run.sh designer`
7. If Mac mocks about “libprism_es2.dylib” cannot be opened because the developer cannot be verified. let him delete it.
8. Go to System Preferences > Security and Allow all those files to be opened
9. Undelete the libs and rerun. This is the tricky part. Take the 2-3 that it deleted from the Recycle Bin and put them back into the lib folder
10. `./run.sh designer`
11. It will still warn of that files but you can now click "Open"
